server/aesdsocket.c

```diff
@@ -158,146 +137,351 @@ int main(int argc, char* argv[])
 		return -1;
 	}
 
+	freeaddrinfo(serverInfo);
+
 	// Run in daemon mode based on flag
 	if(daemonMode)
 	{
 		rc = daemon(-1, -1); // Non-zero values to perform daemon operations
 		if(rc == -1)
 		{
 			perror("daemon");
-			freeaddrinfo(serverInfo);
 			close(serverSockfd);
 			return -1;
 		}
 	}
+
+	// Initialize the thread linked list at this point
+	SLIST_INIT(&head);
 		
 	// Listen for a connection
 	rc = listen(serverSockfd, MAX_PENDING_CONNECTIONS); // max number of pending connections: 3
 	if(rc == -1)
 	{
 		perror("listen");
-		freeaddrinfo(serverInfo);
+		close(serverSockfd);
+		return -1;
+	}
+
+	// Initialize the mutex
+	rc = pthread_mutex_init(&mutex, NULL);
+	if(rc != 0)
+	{
+		perror("pthread_mutex_init");
+		close(serverSockfd);
+		return -1;
+	}
+
+	// Set the timer
+	rc = SetTimer();
+	if(rc == -1)
+	{
 		close(serverSockfd);
 		return -1;
 	}
 	
 	while(!terminate)
 	{
+		// Clean up the list if needed
+		CleanUpList();
+
 		// Accept the connection
-		acceptedSocketfd = accept(serverSockfd, &clientInfo, &clientInfoLen);
-		if(acceptedSocketfd == -1)
+		int acceptedSockFd = accept(serverSockfd, &clientInfo, &clientInfoLen);
+		if(acceptedSockFd == -1)
 		{
 			perror("accept");
-			freeaddrinfo(serverInfo);
 			close(serverSockfd);
 			return -1;
 		}
-		connectionInProgress = true;
-		char clientIpStr[INET6_ADDRSTRLEN];
-		inet_ntop(AF_INET, &clientInfo, clientIpStr, sizeof(clientIpStr));
-		syslog(LOG_INFO, "Accept connection from %s", clientIpStr); // Log client IP to syslog
+		
+		// Create a node for this and add to list
+		struct threadInfo_s* infoPtr = malloc(sizeof(struct threadInfo_s));
+		infoPtr->connectionInProgress = true;
+		infoPtr->sockFd = acceptedSockFd;
+		infoPtr->clientInfo = &clientInfo;
+
+		// Create a thread for the connection
+		rc = pthread_create(&(infoPtr->threadId), NULL, ThreadSendAndReceive, infoPtr);
+		if(rc != 0)
+    	{
+			syslog(LOG_ERR, "Error creating the thread. rc = %d", rc);
+			close(infoPtr->sockFd);
+			close(serverSockfd);
+    		return -1;
+    	}
+		totalConnections++;
```

Comment:
@suryakanteti what happens if a signal occurs just before this line but after `ThreadSendAndReceive` has started?  I think you will end up with multiple timestamps printed in your log file.